### ISSUE TITLE
Using `meta.span_count` in the rules fails when Refinery makes sampling decisions without the root span

### ISSUE THREAD

**Issue by tdarwin** [honeycombio/refinery#850]
Looking at some rules with a customer recently, they have some larger traces with 100k+ spans in a the trace.  However they have the rules below:

```yaml
                - Name: Drop traces with more than 5000 spans
                  Drop: true
                  Conditions:
                    - Field: meta.span_count
                      Operator: '>='
                      Value: 5000
                      Datatype: int
                - Name: sample marketdata trace with more than 1000 spans
                  Conditions:
                    - Field: meta.span_count
                      Operator: '>='
                      Value: 500
                      Datatype: int
                    - Field: service.name
                      Operator: =
                      Value: marketdata
                  Sampler:
                    EMADynamicSampler:
                        GoalSampleRate: 1000
```

A lot of those long traces match these rules, but they're also missing the root span, which explains why they're getting through because the trace doesn't have that field added anywhere.  I think a lot more traffic is getting through that they don't want because this field isn't made available to `trace_send_no_root` traces.

**Versions**

- Refinery: 2.1.0

**Describe the solution you'd like**

I'm thinking that the easiest solution would be to add the `meta.span_count` field to another span should the root span not exist, maybe the first span to arrive at Refinery in that trace? Or if there's another trace that's easily grabbed to add a field to. I think adding it on every span would be unnecessary, but we should make sure it makes it onto _a_ span if the root span doesn't arrive.

**Describe alternatives you've considered**

First Alternative:
We could have a set of pre-caching rules that are evaluated when a span arrives and is processed before going into the cache.  This would allow us to have rules for dropping traces with a certain volume of spans where we could hopefully just look at a count counter for a trace_id real quick and if > threshold, drop the trace and log it in the decision cache. This also has the benefit of being able to handle some traces before the root span comes in _or_ before the trace timeout gets hit.

Second Alternative:
We could have meta-spans that don't get sent, but maintain meta information about a trace, and we can use values in that meta span to drop the `span_count` field in, so that the rule processes based on the data, but the meta span is dropped at the end of the rule processing and not sent upstream to honeycomb.  There's a lot of cool stuff we might be able to do with that concept of a meta span.

**Comment by kentquirk** [comment]
This is basically #646 in a different suit. My thought is to store current span count on the trace object rather than individual spans, so that refinery can act on it at any time. But we also would prefer to recognize when it's too high and execute a drop rule before refinery tries to record a million spans.

**Comment by kentquirk** [comment]
The feature we released in [v2.3.0](https://github.com/honeycombio/refinery/releases/tag/v2.3.0) called [Virtual Fields](https://github.com/honeycombio/refinery/blob/main/rules_conditions.md#virtual-fields) addresses most of the use cases in this issue.

The remaining situation that is hard to handle is:
* A trace doesn't terminate and times out
* But the `?.NUM_DESCENDANTS` value is still less than the rule
* Refinery decides to keep the trace and records a Keep record
* A million other spans arrive for the same trace and get sent to Honeycomb

The general solution would require re-evaluating rules on every span, even late spans, which would be computationally very expensive. It might be possible to address this very specific case with a config value that continues to count late spans and changes the decision after reaching a configured maximum value. But that feels pretty specific right now, so I'm closing this until we see an actual instance of the problem in the field.
